oeps/oep-0037-bp-test-data.rst

```diff
@@ -0,0 +1,254 @@
+load it when needed, and also to keep it working correctly as the code
+changes over time.
+
+This OEP attempts to provide the guidance needed to manage this data
+logically and consistently in order to streamline development, refactoring,
+and automated testing.
+
+Specification
+=============
+
+There are a few different aspects of test data management, each of which
+has different tools that can make it more painless:
+
+Reusing Data
+------------
+
+Once a development or production environment has been populated with a data
+set comprehensive enough to exercise certain functionality in Open edX, it can
+be useful to "save it" for later use.  The first step is to make sure that any
+`PII`_ has been stripped from the data.  This is a particular concern for
+production data that may include customer information, but even development
+data may include a developer's email address and other personal information.
+Tools such as `dj_anonymizer`_ can be useful in making a first pass at
+stripping such PII, but it should be followed by manual verification against
+the `OEP-30`_ annotations.
+
+Once PII has been removed, an attempt can be made to export just the needed
+data.  This may be a combination of course export to OLX and generation of
+fixtures from selected relational data using a utility such as
+`django-fixture-magic`_.  These fixtures should not be saved verbatim, but
+rather used as a guide for producing Python code that generates roughly
+equivalent data using ``factory_boy`` as described below.
+
+Note that if the following recommendations on creating data are followed, test
+data created for a unit test can be fairly easily re-purposed for use as
+development environment data sets.  This can be useful because the starting
+point is already free of PII and has been verified by tests to work as
+intended.
+
+.. _PII: https://open-edx-proposals.readthedocs.io/en/latest/oep-0030-arch-pii-markup-and-auditing.html
+.. _dj_anonymizer: https://github.com/preply/dj_anonymizer
+.. _OEP-30: https://open-edx-proposals.readthedocs.io/en/latest/oep-0030-arch-pii-markup-and-auditing.html
+.. _django-fixture-magic: https://github.com/davedash/django-fixture-magic
+
+Creating Data
+-------------
+
+Whether defining it from scratch or basing it on a dump of data from an
+existing environment, at some point code needs to be written which will
+populate the data stores of each relevant IDA in order to perform the desired
+test or development task.  For relational data in simple unit tests, this is
+often done via the `factory_boy`_ package (and custom model factories using
+it) directly in ``setUpTestData`` or ``setUp``.  For development data (or
+more complex test setup which could be useful to reproduce in a development
+environment), the setup code should instead reside in its own Python function
+or static method.  For example:
+
+.. code-block:: python
+
+    def account_retirement_example(data):
```

Comment:
Should these functions be allowed to take parameters to allow them to be somewhat configurable (for example, how many courses to create?). If so, how could we expose those in the `load_test_data` management command?

Reply:
Yes, the functions could take additional parameters.  I was originally thinking that we'd need custom management commands for such cases, but I guess they could be given as positional arguments instead.  What I originally wrote:

> Most data sets should attempt to use a standard function path in each IDA to
allow use of this generic data loading command, but particularly complex cases
may use custom management commands and/or devstack ``make`` targets (perhaps
even taking arguments to specify parameters of the generated data set).

The devstack make target could probably also take parameters, something like:

```make load_test_data DATA_FUNCTION=test_data.retirement.account_retirement_example DATA_FUNCTION_ARGS=42,second_argument```